### PR TITLE
[Helm][RBAC] Introduce the option crNamespacedRbacEnable to enable or disable the creation of Role/RoleBinding for RayCluster preparation

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

https://github.com/ray-project/kuberay/pull/1106#issuecomment-1578439022

### Without this PR, users have the following options to control the creation of RBAC resources:

(1) Users can set `rbacEnable`: false and `serviceAccount.create: false` to prevent the creation of any RBAC resources.

(2) Users can use the `singleNamespaceInstall` option to determine whether to create namespaced RBAC resources (e.g., Role and RoleBinding) or cluster-scoped RBAC resources (e.g., ClusterRole and ClusterRoleBinding).

(3) If `singleNamespaceInstall` is set to true, users can use the `watchNamespace` option to determine in which namespaces the Role and RoleBinding should be installed. If watchNamespace is not set, the Role and RoleBinding will be installed in the same namespace as the KubeRay operator.

### We can categorize Roles for KubeRay into three categories:
(A) Role for leader election
(B) Role for RayCluster preparation (e.g., Pods, Services) 
(C) ClusterRole for RayCluster preparation (e.g., Pods, Services) 

* If `rbacEnable` is set to false, (A)(B)(C) will not be created.
* The `singleNamespaceInstall` option is used to determine whether to create (B) or (C).

In conclusion, with these two options, users can choose to create nothing, (A)(B), or (A)(C). However, in ArgoCD, a project can only create resources within its associated namespace. Hence, users deploying KubeRay in a Kubernetes cluster managed by GitOps tools would like to only create (A) via Helm chart, and then the cluster admins create Role / RoleBinding for each namespace listed in `watchNamespace` manually later.

To achieve this goal, this PR introduces the option `crNamespacedRbacEnable` to enable or disable the creation of (B). Therefore, ArgoCD users can set `rbacEnable: true`, `singleNamespaceInstall: true`, and `crNamespacedRbacEnable: false` to install only (A).

## Related issue number

#1106 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

* Gist1: [values.yaml](https://gist.github.com/kevin85421/493cb873bc1a53bb09352d8ab552ea85)
  * `rbacEnable: true`, `crNamespacedRbacEnable: false`, `singleNamespaceInstall: true`
* Gist2: [rbac.yaml
](https://gist.github.com/kevin85421/f0eabaf5e60ce44efeaf8d368af8fbea)
  * Role / RoleBinding (i.e. (B)) for `default` namespace

```sh
kind create cluster --image=kindest/node:v1.23.0

# Install the KubeRay operator with Gist1. (path: helm-chart/kuberay-operator)
helm install kuberay-operator .

# Verify that only (A) will be created.
kubectl get role
# NAME                               CREATED AT
# kuberay-operator-leader-election   2023-06-13T07:19:45Z

kubectl get rolebinding
# NAME                               CREATED AT
# kuberay-operator-leader-election   2023-06-13T07:19:45Z

# Check the KubeRay log
kubectl logs $YOUR_OPERATOR_POD

# W0613 07:23:19.622863       1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.23.0/tools/cache/reflector.go:167: failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:kuberay-operator" cannot list resource "pods" in API group "" in the namespace "default": RBAC: clusterrole.rbac.authorization.k8s.io "kuberay-operator" not found
# E0613 07:23:19.623053       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.23.0/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:kuberay-operator" cannot list resource "pods" in API group "" in the namespace "default": RBAC: clusterrole.rbac.authorization.k8s.io "kuberay-operator" not found

# Install Role/RoleBinding for `default` namespace by Gist2
kubectl apply -f rbac.yaml

# Install a RayCluster to verify that the KubeRay operator can create Pods after applying rbac.yaml.
helm install raycluster kuberay/ray-cluster --version 0.5.0
```

### CODE REVIEW

**Comment by kevin85421** [comment]
cc @anshulomar @msumitjain would you mind reviewing this PR? Thanks!

**Review by architkulkarni — APPROVED** [review]
The implementation looks good and the documentation is clear; I'll leave it to ArgoCD folks to verify that the API makes sense and works for the use case.  The only downside is that despite the clear docs, the behavior is inherently kind of complicated, but it sounds like most users won't have to touch this.

File: helm-chart/kuberay-operator/values.yaml [hunk 1]
```diff
@@ -45,18 +45,30 @@ readinessProbe:
...
+rbacEnable: true
+
+# When crNamespacedRbacEnable is set to true, the KubeRay operator will create a Role for RayCluster preparation (e.g., Pods, Services)
+# and a corresponding RoleBinding for each namespace listed in the "watchNamespace" parameter. Please note that even if crNamespacedRbacEnable
+# is set to false, the Role and RoleBinding for leader election will still be created.
+#
+# Note:
+# (1) This variable is only effective when rbacEnable and singleNamespaceInstall are both set to true.
+# (2) In most cases, it should be set to true, unless you are using a Kubernetes cluster managed by GitOps tools such as ArgoCD.
+crNamespacedRbacEnable: true
```
Comment (anshulomar) [inline helm-chart/kuberay-operator/values.yaml:66]: [Minor] I think you might want to rename this field for more clarity. What does "cr" mean: create role?

**Review by anshulomar — APPROVED** [review]

**Comment by anshulomar** [comment]
@msumitjain can you also review this?

File: helm-chart/kuberay-operator/values.yaml (hunk 1 above)
Comment (kevin85421) [reply helm-chart/kuberay-operator/values.yaml:66]: "CR" stands for Custom Resource here. This option is used to control (B) in the PR description. Is `customResourceNamespacedRbacEnabled` a better option?

**Comment by kevin85421** [comment]
Merge this PR. I will open a new PR if @msumitjain has any further feedback.

**Comment by msumitjain** [comment]
Apologies for the late review, thanks @kevin85421  for taking this up. This change looks good to me as well.

**Pull request merged**